llvm/lib/Target/RISCV/RISCVISelLowering.cpp

```diff
@@ -11753,6 +11753,17 @@ SDValue RISCVTargetLowering::lowerVECTOR_INTERLEAVE(SDValue Op,
     return DAG.getMergeValues(Loads, DL);
   }
 
+  // Use ri.vzip2{a,b} if available
+  // TODO: Figure out the best lowering for the spread variants
+  if (Subtarget.hasVendorXRivosVizip() && !Op.getOperand(0).isUndef() &&
+      !Op.getOperand(1).isUndef()) {
```

Comment:
Good catch with the undef case. But should we only skip undefs if `VecVT.getScalarSizeInBits() < Subtarget.getELen()`? At e64 I don't think we get the vzext/vwsll.vx since it gets lowered as a vrgather.vv

Reply:
The vrgather case doesn't have a problem with undef propagation because it's only one use of each operand.  The guard here honestly isn't entirely sufficient, we probably need to be using freeze in both this case and the unzip2a/b case I added.  The guard I added wasn't directly motivated by the legality issue (I'm going to return to that shortly in it's own patch), it was motivated by profitability.  Loosing the zero extend lowering for the spread2 case causes a few of the tests to regress.  I'm going to post a patch to reorganize things, but doing it as a separate change seemed easier.

Reply:
> Loosing the zero extend lowering for the spread2 case causes a few of the tests to regress. 

I wasn't even aware of the legality issue, I just spotted the poison tests and assumed this was so we don't lose the combine. Happy to look at this in a follow up